step-framework-parent/step-framework-collections/src/main/java/step/core/ql/OQLAttributesAwareFilterVisitor.java

```diff
@@ -0,0 +1,146 @@
+package step.core.ql;
+
+import step.core.collections.Filter;
+import step.core.collections.Filters;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class OQLAttributesAwareFilterVisitor extends OQLBaseVisitor<Filter> {
```

Comment:
@jeromecomte that was initial implementation, but you guys said "new visitor / builder, no direct change to the existing OQLFilterBuilder class". If you are ok, I prefer to extend the existing one

Reply:
@lucianbaciu We'd like to avoid having TS specific code in the generic OQL layer and therefore thought about extending the class in the java sense. If this isn't possible you could implement the tracking of fields in a generic way in the existing class

Reply:
@lucianbaciu I also had a look and to me the class can extends OQLFilterVisitor removing the need of duplicated code, let us know otherwise. Also please use a private method for the duplicated code:
        if (ignoreAttributes.contains(text0)) {
            return Filters.empty();
        }
        attributes.add(text0);

Reply:
alright!

Reply:
@jeromecomte @david-stephan after some research on these little classes, this is the idea:
There are 3 new things which the visitor has to be aware of:
1. custom processing of keys (can be easily done via a 'protected proceessKey()' method, each class being able to extend it and apply its own custom logic.
2. ignoring of specific attributes - here is a bit more complicated, because this is not just processing, but I need to break inside the existing functions - returning an empty filter.
3. collecting of keys - not an issue, it can be done directly in the original filter.

I see a few options to solve this:
1. Duplicating code - I also don't like it at all, especially for ugly low level processing code.
2. Make the original visitor more generic with a ProcessParams object in the constructor (optional). Here we could pass transforming functions for both keys and values, attributes to be ignored, or any other later on). In this case the methods have to be adjusted, to handle these optional settings. This would be the well known 'interceptors pattern'.
3. Another approach (I see its value on the long term) is to create a RAW structure of the OQL while processing it. This means together with the final immutable Filter, we build another structure with an Operation object for each (key, operator, and value - or left, operator, right). Having this would allow us to use OQL in many other places, because we would have control on it. With a simple operation we could get all the keys from the structure, and process it outside of this layer).

If you find better/simpler approaches, please let me know. 
P.S: I still see it problematic that our generated Filter is not even read-only - 0 extension capabilities, so my vote would go for 3.

Reply:
@lucianbaciu Thanks for the propositions. Agree with you that 3. would be the better solution on the long term. We could create a ticket for it. However in order to minimize the effort on this I propose to extend the class as discussed. I've implemented a proposition on following branch: github.com/exense/step-framework/commit/ab8076069ea295fc8cf229de1fd8f2efd02f6cf9 It passed the existing Unit tests however as mentioned in the comment we don't have full test coverage of this class. Could you therefore review these changes, increase the coverage and merge if it works as intended?

Reply:
@jeromecomte yes, sure!